src/DotNetWorker.Grpc/Handlers/InvocationHandler.cs

```diff
@@ -0,0 +1,159 @@
+                invocationFeatures.Set<IFunctionBindingsFeature>(new GrpcFunctionBindingsFeature(context, request, _outputBindingsInfoProvider));
+
+                if (_inputConversionFeatureProvider.TryCreate(typeof(DefaultInputConversionFeature), out var conversion))
+                {
+                    invocationFeatures.Set<IInputConversionFeature>(conversion!);
+                }
+
+                await _application.InvokeFunctionAsync(context);
+
+                var functionBindings = context.GetBindings();
+
+                foreach (var binding in functionBindings.OutputBindingData)
+                {
+                    var parameterBinding = new ParameterBinding
+                    {
+                        Name = binding.Key
+                    };
+
+                    if (binding.Value is not null)
+                    {
+                        parameterBinding.Data = await binding.Value.ToRpcAsync(_serializer);
+                    }
+
+                    response.OutputData.Add(parameterBinding);
+                }
+
+                if (functionBindings.InvocationResult is not null)
+                {
+                    TypedData? returnVal = await functionBindings.InvocationResult.ToRpcAsync(_serializer);
+                    response.ReturnValue = returnVal;
+                }
+
+                response.Result.Status = StatusResult.Types.Status.Success;
+            }
+            catch (Exception ex)
+            {
+                response.Result.Exception = ex.ToRpcException();
+                response.Result.Status = StatusResult.Types.Status.Failure;
+
+                if (ex.InnerException is TaskCanceledException or OperationCanceledException)
+                {
+                    response.Result.Status = StatusResult.Types.Status.Cancelled;
+                }
+            }
+            finally
+            {
+                _inflightInvocations.TryRemove(request.InvocationId, out var cts);
+
+                if (context is IAsyncDisposable asyncContext)
+                {
+                    await asyncContext.DisposeAsync();
+                }
+
+                (context as IDisposable)?.Dispose();
+            }
+
+            return response;
+        }
+
+        public bool TryCancel(string invocationId)
+        {
+            _inflightInvocations.TryGetValue(invocationId, out var cancellationTokenSource);
+
+            if (cancellationTokenSource is CancellationTokenSource)
+            {
+                try
+                {
+                    cancellationTokenSource?.Cancel();
+                    return true;
+                }
+                catch (ObjectDisposedException)
+                {
+                    // Do nothing, normal behavior
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogWarning($"Unable to cancel invocation {invocationId}.", ex);
```

Comment:
With `ILogger`, pass exception as the first argument (exceptions have special logic) and use message templates (_not_ string interpolation). The logger will actually throw here because you are supplying 1 template value, but have 0 holes to fill.
```suggestion
                    _logger.LogWarning(ex, "Unable to cancel invocation {invocationId}.", invocationId);
```

Reply:
Good to know, will keep this in mind :D

Reply:
We may also want to consider high performance logging - but that can be done in another PR.

Essentially, logging like this requires the message template to be parsed each time. High performance lets us pre-define these log messages that are generated at compile time.

https://docs.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator

Reply:
Ooo very cool, that'd definitely sounds like an improvement worth making! Feel free to create an issue for this and we can get it triaged in the next backlog grooming meeting